common/src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -250,6 +250,80 @@ class SchemaService(private val options: Neo4jOptions, private val driverCache:
     StructType(sortedStructFields)
   }
 
+  def structForGDS(): StructType = {
+    val query =
+      """
+        |CALL gds.list() YIELD name, signature, type
+        |WHERE name = $procName AND type = 'procedure'
+        |WITH split(signature, ') :: (')[1] AS fields
+        |WITH substring(fields, 0, size(fields) - 1) AS fields
+        |WITH split(fields, ',') AS fields
+        |WITH [field IN fields | split(field, ' :: ')] AS fields
+        |UNWIND fields AS field
+        |WITH field
+        |RETURN *
+        |""".stripMargin
+    val map: util.Map[String, AnyRef] = Map[String, AnyRef]("procName" -> options.query.value).asJava
+    val fields = session.run(query, map).list.asScala
+      .map(r => r.get("field").asList((t: Value) => t.asString()).asScala)
+      .map(r => (r.head.trim, r(1).replaceAll("\\?", "") match {
+        case "STRING" => ("String", null)
+        case "INTEGER" => ("Long", null)
+        case "FLOAT" | "NUMBER" => ("Double", null)
+        case "DATETIME" => ("DateTime", null)
+        case "BOOLEAN" => ("Boolean", null)
+        case "LOCALTIME" => ("LocalTime", null)
+        case "LIST OF INTEGER" => ("LongArray", null)
+        case "LIST OF FLOAT" => ("DoubleArray", null)
+        case "LIST OF FLOAT" => ("DoubleArray", null)
+        case "LIST OF STRING" => ("StringArray", null)
+        case "MAP" =>
+          logWarning(
+            s"""
+               |For procedure ${options.query.value}
+               |Neo4j return type MAP? of field ${r.head.trim} not fully supported.
+               |We'll coerce it to a Map<String, String>
+               |""".stripMargin)
+          ("Map", Map("key" -> "").asJava) // dummy value
+        case "LIST OF MAP" =>
+          logWarning(
+            s"""
+               |For procedure ${options.query.value}
+               |Neo4j return type LIST? OF MAP? of field ${r.head.trim} not fully supported.
+               |We'll coerce it to a [Map<String, String>]
+               |""".stripMargin)
+          ("MapArray", Seq(Map("key" -> "").asJava).asJava) // dummy value
+        case "PATH" => ("Path", null)
+        case _ => throw new IllegalArgumentException(s"Neo4j type ${r(1)} not supported")
+      }))
+      .map(r => StructField(r._1, cypherToSparkType(r._2._1, r._2._2)))
+      .toSeq
+    StructType(fields)
+  }
+
+  def inputForGDSProc(procName: String): Seq[(String, Boolean)]  = {
```

Comment:
probably good enough for now, but maybe it's worth caching the whole proc list information at startup so you don't have to run this every time you want to run a GDS proc?

Reply:
It's invoked just once per spark job.